### DISCUSSION TITLE
Retrieving auth state using Pocketbase as a framework

### DISCUSSION THREAD

**Discussion by ckrowbar** [pocketbase/pocketbase#4034] [Q&A]
Right now, I'm trying to extend Pocketbase to have routes that allow me to return info from a user's Spotify account. In doing so, I want to verify that the get request is only coming from the user, and not just someone who happens to know that users ID. To do this, I saw that Pocketbase has special keys that can show the authState of a given context, [as seen here](https://pocketbase.io/docs/go-routing/#retrieving-the-current-auth-state).

However, when I tried copy-pasting the code from the docs, I am getting nils instead of the given user (though other info fields, such as browser info, are populated). This is the case for both `c.Get(apis.ContextAuthRecordKey).(*models.Record)` and `apis.RequestInfo(c).authRecord`. I had 3 questions:

1. How is the context object getting the info for the authenticated user? I'm guessing this data is in the header but was unsure why auth state would be there

2. Am I tackling this problem of authentication the right way? Or is there some better convention.

3. Any pointers as to why the user authState is always nil? I can verify that I'm correctly signed in on the client side, and that other info is being passed in the apis.RequestInfo(c), so it's not immediately obvious if my issue stems from implementation or understanding.

Thanks in advanced, and happy holidays!

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.

In order `c.Get(apis.ContextAuthRecordKey)` to return a non-nil value you need to be authenticated before sending the request to your custom route, aka. the request to have `Authorization: TOKEN` header with it. This header is added automatically by the SDKs when `pb.authStore` is populated.

Note that PocketBase APIs are fully stateless, they don't have sessions in the traditional sense and each request acts independently from one another.

If you are accessing your custom route by directly opening its url in the browser, then the auth state is expected to be nil because the request will not have `Authorization` header with it.

If you want to emulate session-like behavior then you can use cookies to store the token and populate the auth state from the cookie with a middleware. It is a little outdated but you can check for an example https://github.com/pocketbase/pocketbase/discussions/989.

If the above doesn't help, please provide a code sample with your custom action and elaborate how you are sending the request to it.

**Reply by ckrowbar** [reply]
Thanks for the thorough reply. 

Right now I'm using the SDKs, and failing to get the authentication in the response. I would also imagine that "this header is added automatically by the SDKs when pb.authStore is populated", though I can't seem to make it work. Here is some code that I'm runing (svelte frontend/pb backend running locally) that is failing to do what I expect it to despite Spotify auth working. 

In my Svelte frontend:
```
<script>
    import { user, pb, updateUser } from './pocketbase';

    async function login() {
        const authData = await pb.collection('users').authWithOAuth2({ provider: 'spotify', scopes: ['playlist-read-private'] });
        const data = {
            "accessToken": authData.meta.accessToken,
            "refreshToken": authData.meta.refreshToken,
            "name": authData.meta.name,
            "avatarUrl": authData.meta.rawUser.images[1].url,
        }
        updateUser($user.id, data)
        console.log(authData)
    };

    async function fetchPlaylists() {
        console.log('fetching playlists');
        console.log(pb.authStore);
        const response = await fetch(`http://127.0.0.1:8090/api/playlists`);
        const data = await response.json();
        console.log(data);
    }
    
    function logout() {
        pb.authStore.clear();
    }
</script>

{#if $user}
    <p>Signed in as {$user.name}</p>
    <button class="button" on:click={logout}>Logout</button>
    <button class="button" on:click={fetchPlaylists}>Fetch Playlists</button>
{:else}
    <p>Login/signup below</p>
    <button class="button" on:click={login}>Login</button>
{/if} 
```

In my PB backend:
```
// main.go
package main

import (
	... // removed for formatting, proper imports are here
)

func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    e.Router.GET("/api/playlists", 
    func(c echo.Context) error {
	    playlists := []map[string]string{}
	    
	    record, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
	    
	    if record != nil {
		    playlists = append(playlists, map[string]string{
			    "name": "test",
		    })
	    }
    
	    return c.JSON(200, playlists)
    })
    
    return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

One would expect that clicking the fetch playlist buttons would return a non-empty list when the user is authenticated. However, instead we see the following logs.  

![image](https://github.com/pocketbase/pocketbase/assets/62040387/d7dc2498-6ece-4815-b475-a9b90fe8659b)

I initially thought I was not passing the auth properly, but if the SDK handles this there might be some other issue with the code. I can also confirm that the Spotify auth is working as intended as I can pull the appropriate data from the user. Let me know if there's any glaring issues here, or if I'm misunderstanding how to make my auth state known.

**Reply by ganigeorgiev** [reply]
You are sending the `/api/playlists` request with a plain `fetch` and not the SDK, so the auth header will not be added.

For sending requests with the SDK to custom routes use the `send()` method or manually add the `Authorization: YOUR_TOKEN` header to the `fetch` request.

```js
const result = await pb.send("/api/playlists", { /* send options */ });
```
